File: APKInstaller/src/MultiPackageDialog.vb
Imports System.IO
Imports APKInstaller.My.Resources
Imports MaterialSkin

Public Class MultiPackageDialog
    Private ReadOnly _files As String()
    Private _modifying As Boolean

    Private Sub New(files As String())

        ' This call is required by the designer.
        InitializeComponent()

        ' Add any initialization after the InitializeComponent() call.
        _files = files
    End Sub


    Public Shared Function Create(files As String()) As MultiPackageDialog
        Return New MultiPackageDialog(files)
    End Function

    Private Sub MultiPackageDialog_Load(sender As Object, e As EventArgs) Handles Me.Load
        lstFiles.Items.AddRange(_files.ToArray)

        'Configure GUI
        Dim manager = MaterialSkinManager.Instance
        manager.AddFormToManage(Me)
        'SkinManager.Theme = MaterialSkinManager.Themes.LIGHT
        'S 'kinManager.ColorScheme = New ColorScheme(Primary.Red800, Primary.Red800, Primary.Red200, Accent.LightBlue200, TextShade.WHITE)
        CenterToScreen()

        btnDelete.Enabled = False
        btnModify.Enabled = False
        lstFiles.DrawMode = DrawMode.OwnerDrawFixed
    End Sub

    Function GetFiles() As String()
        Dim list(lstFiles.Items.Count) As String
        lstFiles.Items.CopyTo(list, 0)
        Return list
    End Function

    Private Sub btnOk_Click(sender As Object, e As EventArgs) Handles btnOk.Click
        For Each item In lstFiles.Items
            If Not Installer.ValidateFile(item.ToString(), True) Then
                If MsgBox("An invalid or nonexistent APK file was found. If you continue, the file may not be installed. Do you want to continue?", CType(MsgBoxStyle.YesNo + MsgBoxStyle.Exclamation, Global.Microsoft.VisualBasic.MsgBoxStyle)) = MsgBoxResult.No Then
                    Exit Sub
                End If
            End If
        Next
        DialogResult = DialogResult.OK
        Close()
    End Sub

    Private Sub btnCancel_Click(sender As Object, e As EventArgs) Handles btnCancel.Click
        If MsgBox("Are you really sure you want to cancel? None of the changes made will be preserved.", CType(MsgBoxStyle.Exclamation + MsgBoxStyle.YesNo, Global.Microsoft.VisualBasic.MsgBoxStyle)) = MsgBoxResult.Yes Then
            DialogResult = DialogResult.Cancel
            Close()
        End If
    End Sub

    Private Sub lstFiles_SelectedIndexChanged(sender As Object, e As EventArgs) Handles lstFiles.SelectedIndexChanged
        If lstFiles.SelectedIndex >= 0 Then
            btnModify.Enabled = True
            btnDelete.Enabled = True
        Else
            btnModify.Enabled = False
            btnDelete.Enabled = False
        End If

        'modifying = True
    End Sub

    Private Sub btnAdd_Click(sender As Object, e As EventArgs) Handles btnAdd.Click
        If _modifying And lstFiles.SelectedIndex >= 0 Then

            'btnAdd.Text = "Add"
            'modifying = False
            'tnBrowse.Visible = True
        End If

        lstFiles.Items.Add(txtFile.Text)
        lstFiles.SelectedIndex = lstFiles.Items.Count - 1
        lstFiles.Enabled = True

        txtFile.Text = ""
    End Sub

    Private Sub btnDelete_Click(sender As Object, e As EventArgs) Handles btnDelete.Click
        lstFiles.Items.RemoveAt(lstFiles.SelectedIndex)
    End Sub


    Private Sub listBox1_DrawItem(sender As Object, e As DrawItemEventArgs) Handles lstFiles.DrawItem
        e.DrawBackground()

        Dim i = 0
        Dim g = lstFiles.CreateGraphics()
        For Each item In lstFiles.Items
            Dim sizeF = g.MeasureString(item.ToString, lstFiles.Font)
            If sizeF.Width > i Then
                i = CInt(sizeF.Width)
            End If
        Next
        lstFiles.HorizontalExtent = i

        Using b As New SolidBrush(e.ForeColor)
            If e.Index >= 0 Then
                Dim itemText = lstFiles.GetItemText(lstFiles.Items(e.Index))
                'If Not Installer.ValidateFile(itemText) Then
                '    e.Graphics.FillRectangle(Brushes.OrangeRed, e.Bounds)
                'End If
                If (e.State And DrawItemState.Selected) = DrawItemState.Selected Then
                    e.Graphics.FillRectangle(MaterialSkinManager.Instance.ColorScheme.AccentBrush, e.Bounds)
                End If
                e.Graphics.DrawString(itemText, e.Font, b, e.Bounds)
                End If
        End Using

        e.DrawFocusRectangle()
    End Sub

    Private Sub btnModify_Click(sender As Object, e As EventArgs) Handles btnModify.Click
        If lstFiles.SelectedIndex >= 0 Then
            If _modifying Then
                lstFiles.Items.RemoveAt(lstFiles.SelectedIndex)
                lstFiles.Items.Add(txtFile.Text)
                lstFiles.SelectedIndex = lstFiles.Items.Count - 1

                lstFiles.Enabled = True
                btnModify.Text = Strings.Modify
                _modifying = False
            Else
                txtFile.Text = lstFiles.SelectedItem.ToString
                lstFiles.Enabled = False
                btnModify.Text = Strings.Confirm
                _modifying = True
            End If
        End If
    End Sub

    Private Sub btnBrowse_Click(sender As Object, e As EventArgs) Handles btnBrowse.Click
        Using fileDialog = New OpenFileDialog()

            fileDialog.AutoUpgradeEnabled = True
            fileDialog.CheckFileExists = True
            fileDialog.DefaultExt = ".apk"
            fileDialog.Title = Strings.openFileDialogTitle
            fileDialog.Multiselect = False
            fileDialog.ValidateNames = True
            fileDialog.Filter = Strings.openFileDialogFilter
            fileDialog.ShowDialog()

            txtFile.Text = fileDialog.FileName
        End Using
    End Sub

    Private Sub txtFile_TextChanged(sender As Object, e As EventArgs) Handles txtFile.TextChanged
        Dim validateFile = Installer.ValidateFile(txtFile.Text)
        btnAdd.Enabled = validateFile
        If _modifying Then
            btnModify.Enabled = validateFile
        End If
    End Sub
End Class

File: AppInstaller/MultiPackageDialog.cs
using System;
using System.Drawing;
using System.Linq;
using System.Windows.Forms;
using APKInstaller.i18n;
//using Microsoft.VisualBasic;

namespace APKInstaller
{
    public partial class MultiPackageDialog : Form
    {
        private readonly string[] _files;
        private bool _modifying;

        public MultiPackageDialog(string[] files)
        {
            InitializeComponent();
            _files = files;
        }



        public static MultiPackageDialog Create(string[] files)
        {
            return new MultiPackageDialog(files);
        }

        private void MultiPackageDialog_Load(object sender, EventArgs e)
        {
            lstFiles.Items.AddRange(_files.ToArray());

            //Configure GUI
            //Dim manager = MaterialSkinManager.Instance
            //manager.AddFormToManage(Me)
            //SkinManager.Theme = MaterialSkinManager.Themes.LIGHT
            //S 'kinManager.ColorScheme = New ColorScheme(Primary.Red800, Primary.Red800, Primary.Red200, Accent.LightBlue200, TextShade.WHITE)
            CenterToScreen();

            btnDelete.Enabled = false;
            btnModify.Enabled = false;
            lstFiles.DrawMode = DrawMode.OwnerDrawFixed;
        }

        public string[] GetFiles()
        {
            string[] list = new string[lstFiles.Items.Count + 1];
            lstFiles.Items.CopyTo(list, 0);
            return list;
        }

        private void btnOk_Click(object sender, EventArgs e)
        {
            DialogResult = DialogResult.OK;
            Close();
        }

        private void btnCancel_Click(object sender, EventArgs e)
        {
            if (MessageBox.Show("Are you really sure you want to cancel? None of the changes made will be preserved.", "Confirm Cancel", MessageBoxButtons.YesNo, MessageBoxIcon.Warning) == DialogResult.Yes)
            {
                DialogResult = DialogResult.Cancel;
                Close();
            }
        }

        private void lstFiles_SelectedIndexChanged(object sender, EventArgs e)
        {
            if (lstFiles.SelectedIndex >= 0)
            {
                btnModify.Enabled = true;
                btnDelete.Enabled = true;
            }
            else
            {
                btnModify.Enabled = false;
                btnDelete.Enabled = false;
            }

            //modifying = True
        }

        private void btnAdd_Click(object sender, EventArgs e)
        {

            if (_modifying & lstFiles.SelectedIndex >= 0)
            {
                //btnAdd.Text = "Add"
                //modifying = False
                //tnBrowse.Visible = True
            }

            lstFiles.Items.Add(txtFile.Text);
            lstFiles.SelectedIndex = lstFiles.Items.Count - 1;
            lstFiles.Enabled = true;

            txtFile.Text = "";
        }

        private void btnDelete_Click(object sender, EventArgs e)
        {
            lstFiles.Items.RemoveAt(lstFiles.SelectedIndex);
        }


        private void listBox1_DrawItem(object sender, DrawItemEventArgs e)
        {
            e.DrawBackground();

            dynamic i = 0;
            dynamic g = lstFiles.CreateGraphics();
            foreach (var item in lstFiles.Items)
            {
                //item = item_loopVariable;
                dynamic sizeF = g.MeasureString(item.ToString(), lstFiles.Font);
                if (sizeF.Width > i)
                {
                    i = Convert.ToInt32(sizeF.Width);
                }
            }
            lstFiles.HorizontalExtent = i;

            using (SolidBrush b = new SolidBrush(e.ForeColor))
            {
                if (e.Index >= 0)
                {
                    dynamic itemText = lstFiles.GetItemText(lstFiles.Items[e.Index]);
                    //If Not Installer.ValidateFile(itemText) Then
                    //    e.Graphics.FillRectangle(Brushes.OrangeRed, e.Bounds)
                    //End If
                    if ((e.State & DrawItemState.Selected) == DrawItemState.Selected)
                    {
                        e.Graphics.FillRectangle(Brushes.Aqua, e.Bounds);
                    }
                    e.Graphics.DrawString(itemText, e.Font, b, e.Bounds);
                }
            }

            e.DrawFocusRectangle();
        }

        private void btnModify_Click(object sender, EventArgs e)
        {
            if (lstFiles.SelectedIndex >= 0)
            {
                if (_modifying)
                {
                    lstFiles.Items.RemoveAt(lstFiles.SelectedIndex);
                    lstFiles.Items.Add(txtFile.Text);
                    lstFiles.SelectedIndex = lstFiles.Items.Count - 1;

                    lstFiles.Enabled = true;
                    btnModify.Text = UIStrings.Modify;
                    _modifying = false;
                }
                else
                {
                    txtFile.Text = lstFiles.SelectedItem.ToString();
                    lstFiles.Enabled = false;
                    btnModify.Text = UIStrings.Confirm;
                    _modifying = true;
                }
            }
        }

        private void btnBrowse_Click(object sender, EventArgs e)
        {
            using (var fileDialog = new OpenFileDialog())
            {

                fileDialog.AutoUpgradeEnabled = true;
                fileDialog.CheckFileExists = true;
                fileDialog.DefaultExt = ".apk";
                fileDialog.Title = UIStrings.openFileDialogTitle;
                fileDialog.Multiselect = false;
                fileDialog.ValidateNames = true;
                fileDialog.Filter = UIStrings.openFileDialogFilter;
                fileDialog.ShowDialog();

                txtFile.Text = fileDialog.FileName;
            }
        }

        private void txtFile_TextChanged(object sender, EventArgs e)
        {
            dynamic validateFile = Installer.ValidateFile(txtFile.Text);
            btnAdd.Enabled = validateFile;
            if (_modifying)
            {
                btnModify.Enabled = validateFile;
            }
        }

        private void MultiPackageDialog_Load_1(object sender, EventArgs e)
        {

        }

        //=======================================================
        //Service provided by Telerik (www.telerik.com)
        //Conversion powered by NRefactory.
        //Twitter: @telerik
        //Facebook: facebook.com/telerik
        //=======================================================

    }
}
